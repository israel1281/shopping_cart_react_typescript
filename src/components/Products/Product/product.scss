.productContainer {
  position: relative;
  text-align: center;
  box-sizing: border-box;
  padding: 10px;
  margin-bottom: 30px;
  cursor: default;
  outline: none;

  .shopping {
    position: absolute;
    color: #ececec;
    top: 10px;
    right: 10px;
    padding: 5px;
    font-size: 0.6em;
    background-color: #001f2f;
    border-bottom-left-radius: 4px;
    cursor: default;
    z-index: 1;
  }

  .productImg {
    width: 100%;
    height: 270px;
    position: relative;
    object-fit: cover;
    transition: 0.2s;
    border-radius: 4px;

    @media only screen and (min-width: 768px) {
      height: 320px;
    }
  }

  .productTitle {
    position: relative;
    padding: 0 20px;
    height: 48px;
    font-weight: 500;

    &::before {
      content: "";
      width: 20px;
      height: 2px;
      background-color: #001f2f;
      position: absolute;
      bottom: 0;
      left: 50%;
      margin-left: -10px;
    }
  }

  .productPrice {
    color: #001f2f;
    height: 60px;
    .productValue {
      margin: 0;
      b {
        color: #001f2f;
        font-size: 1.5em;
        margin-left: 5px;
      }
    }

    .installment {
      margin: 0;
      color: #9c9b9b;
    }
  }

  .buyButton {
    background-color: #f2ab06ef;
    color: #fff;
    padding: 15px 0;
    margin-top: 10px;
    font-size: 0.9rem;
    cursor: pointer;
    font-weight: 500;
    width: 100%;
    border: 0;
    border-radius: 4px;
    transition: background-color 0.2s;

    &:hover {
      background-color: rgb(0, 44, 65);
    }
  }
}
