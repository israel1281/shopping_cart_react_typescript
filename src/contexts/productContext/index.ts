export { ProductsProvider } from "./ProductContextProvider";
export { default as useProducts } from "./useProducts";
