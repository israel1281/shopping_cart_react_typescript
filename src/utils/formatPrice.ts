const formatPrice = (price: number, currencyId: string): string => {
  switch (currencyId) {
    case "NGN":
      return price.toFixed(4).replace(".", ",");
    default:
      return price.toFixed(2);
  }
};

export default formatPrice;
