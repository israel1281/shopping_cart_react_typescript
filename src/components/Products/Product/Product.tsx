import { KeyboardEvent, useState } from "react";

import formatPrice from "../../../utils/formatPrice";
import { IProduct } from "../../../models";

import { useCart } from "../../../contexts/cartContext";
import "./product.scss";

interface IProps {
  product: IProduct;
}

const Product = ({ product }: IProps) => {
  const { openCart, addProduct } = useCart();
  const {
    sku,
    title,
    price,
    installments,
    currencyId,
    currencyFormat,
    isFreeShipping,
  } = product;

  const [hover, setHover] = useState(false);

  const formattedPrice = formatPrice(price, currencyId);
  let productInstallment;

  if (installments) {
    const installmentPrice = price / installments;

    productInstallment = (
      <p className="installment">
        <span>or {installments} x</span>
        <b>
          {currencyFormat}
          {formatPrice(installmentPrice, currencyId)}
        </b>
      </p>
    );
  }

  const handleAddProduct = () => {
    addProduct({ ...product, quantity: 1 });
    openCart();
  };

  const handleAddProductWhenEnter = (event: KeyboardEvent) => {
    if (event.key === "Enter" || event.code === "Space") {
      addProduct({ ...product, quantity: 1 });
      openCart();
    }
  };

  return (
    <div
      className="productContainer"
      onKeyUp={handleAddProductWhenEnter}
      key={sku}
      tabIndex={1}
    >
      {isFreeShipping && <div className="shopping">Free shipping</div>}
      <img
        className="productImg"
        src={
          hover
            ? require(`../../../assets/products/${sku}-2-product.webp`)
            : require(`../../../assets/products/${sku}-1-product.webp`)
        }
        alt={title}
        onPointerOver={() => setHover(true)}
        onPointerOut={() => setHover(false)}
      />
      <p className="productTitle">{title}</p>
      <div className="productPrice">
        <p className="productValue">
          <small>{currencyFormat}</small>
          <b>{formattedPrice.substring(0, formattedPrice.length - 3)}</b>
          <span>{formattedPrice.substring(formattedPrice.length - 3)}</span>
        </p>
        {productInstallment}
      </div>
      <button className="buyButton" onClick={handleAddProduct} tabIndex={-1}>
        Add to cart
      </button>
    </div>
  );
};

export default Product;
