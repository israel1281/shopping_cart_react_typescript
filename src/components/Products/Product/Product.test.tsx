import { CartProvider } from "../../../contexts/cartContext";
import { mockCartProducts } from "../../../utils/mockData";

import Product from ".";

describe("Product Component", () => {
  const setup = (props = {}) => {
    return (
      <CartProvider>
        <Product product={mockCartProducts[0]} {...props} />
      </CartProvider>
    );
  };

  test("should render correctly", () => {
    const view = setup();
    expect(view).toMatchSnapshot();
  });
});
