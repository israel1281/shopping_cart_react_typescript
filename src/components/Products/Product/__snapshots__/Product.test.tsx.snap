// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[components] - Product should render correctly 1`] = `
<CartProvider>
  <Product
    product={
      Object {
        "availableSizes": Array [
          "S",
          "L",
          "XL",
          "XXL",
        ],
        "currencyFormat": "$",
        "currencyId": "USD",
        "description": "14/15 s/nº",
        "id": 0,
        "installments": 9,
        "isFreeShipping": true,
        "price": 10.9,
        "quantity": 1,
        "sku": 8552515751438644,
        "style": "White T-shirt",
        "title": "Cropped Stay Groovy off white",
      }
    }
  />
</CartProvider>
`;

exports[`Product Component should render correctly 1`] = `
<CartProvider>
  <Product
    product={
      Object {
        "availableSizes": Array [
          "S",
          "L",
          "XL",
          "XXL",
        ],
        "currencyFormat": "$",
        "currencyId": "USD",
        "description": "14/15 s/nº",
        "id": 0,
        "installments": 9,
        "isFreeShipping": true,
        "price": 10.9,
        "quantity": 1,
        "sku": 8552515751438644,
        "style": "White T-shirt",
        "title": "Cropped Stay Groovy off white",
      }
    }
  />
</CartProvider>
`;
